## Week 11, Lab 1: SQL Practice

> Author: Matt Brems

In this lab, you're going to have ten prompts. Each prompt will require you to do two things:
1. Write a SQL query.
2. Use the result of that query to answer the question.

The purpose of this lab is to get you to practice your SQL - an **important** skill in data science! While it is possible to get these answers using Pandas, it should be your goal to do everything only in SQL.

**Data**: The data used in this lab is the famous (but fake) Northwind database. It contains various tables of transactions. The schema for the Northwind database can be found [here](http://www.zentut.com/wp-content/uploads/downloads/2013/06/Northwind-Sample-Database-Diagram.pdf).

In [1]:
# Import Pandas and Create_Engine

import pandas as pd
from sqlalchemy import create_engine
engine = create_engine('postgres://jlpmxjkftcsepc:bb9cd75e6659d7d67ee3b86be0f905069319c75bd6823d786c808dd799ea85a2@ec2-54-227-251-33.compute-1.amazonaws.com:5432/d9teltkv30dni7')

### Prompt 1

How many rows are there in the `customers` table of the Northwind database?

In [2]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT *
FROM customers
"""

display(pd.read_sql_query(sql, engine))
engine.dispose()

customer_id                          company_name             contact_name  \
0        ALFKI                   Alfreds Futterkiste             Maria Anders   
1        ANATR    Ana Trujillo Emparedados y helados             Ana Trujillo   
2        ANTON               Antonio Moreno Taquería           Antonio Moreno   
3        AROUT                       Around the Horn             Thomas Hardy   
4        BERGS                    Berglunds snabbköp       Christina Berglund   
5        BLAUS               Blauer See Delikatessen               Hanna Moos   
6        BLONP              Blondesddsl père et fils       Frédérique Citeaux   
7        BOLID             Bólido Comidas preparadas            Martín Sommer   
8        BONAP                              Bon app'         Laurence Lebihan   
9        BOTTM                 Bottom-Dollar Markets        Elizabeth Lincoln   
10       BSBEV                         B's Beverages        Victoria Ashworth   
11       CACTU            Cactus Comidas para llevar         Patricio Simpson   
12       CENTC            Centro comercial Moctezuma          Francisco Chang   
13       CHOPS                     Chop-suey Chinese                Yang Wang   
14       COMMI                      Comércio Mineiro             Pedro Afonso   
15       CONSH                 Consolidated Holdings          Elizabeth Brown   
16       DRACD              Drachenblut Delikatessen             Sven Ottlieb   
17       DUMON                       Du monde entier           Janine Labrune   
18       EASTC                    Eastern Connection                Ann Devon   
19       ERNSH                          Ernst Handel            Roland Mendel   
20       FAMIA                    Familia Arquibaldo                Aria Cruz   
21       FISSA  FISSA Fabrica Inter. Salchichas S.A.               Diego Roel   
22       FOLIG                     Folies gourmandes            Martine Rancé   
23       FOLKO                        Folk och fä HB            Maria Larsson   
24       FRANK                        Frankenversand            Peter Franken   
25       FRANR                   France restauration           Carine Schmitt   
26       FRANS                        Franchi S.p.A.            Paolo Accorti   
27       FURIB        Furia Bacalhau e Frutos do Mar           Lino Rodriguez   
28       GALED                Galería del gastrónomo         Eduardo Saavedra   
29       GODOS                   Godos Cocina Típica        José Pedro Freyre   
..         ...                                   ...                      ...   
61       QUEEN                         Queen Cozinha           Lúcia Carvalho   
62       QUICK                            QUICK-Stop              Horst Kloss   
63       RANCH                         Rancho grande         Sergio Gutiérrez   
64       RATTC            Rattlesnake Canyon Grocery             Paula Wilson   
65       REGGC                    Reggiani Caseifici          Maurizio Moroni   
66       RICAR                    Ricardo Adocicados           Janete Limeira   
67       RICSU                    Richter Supermarkt             Michael Holz   
68       ROMEY                      Romero y tomillo         Alejandra Camino   
69       SANTG                         Santé Gourmet         Jonas Bergulfsen   
70       SAVEA                    Save-a-lot Markets           Jose Pavarotti   
71       SEVES                    Seven Seas Imports               Hari Kumar   
72       SIMOB                         Simons bistro           Jytte Petersen   
73       SPECD                  Spécialités du monde        Dominique Perrier   
74       SPLIR                 Split Rail Beer & Ale       Art Braunschweiger   
75       SUPRD                      Suprêmes délices         Pascale Cartrain   
76       THEBI                        The Big Cheese                Liz Nixon   
77       THECR                       The Cracker Box                 Liu Wong   
78       TOMSP                    

**Prompt 1 Answer**:

91 Rows

### Prompt 2

If you sort the customers in alphabetical order by name, which customer is fourth?

In [3]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT c.contact_name
FROM customers AS c
ORDER BY contact_name
"""

df = pd.read_sql_query(sql, engine)
df.head()

contact_name
0   Alejandra Camino
1    Alexander Feuer
2  Anabela Domingues
3       Ana Trujillo
4      André Fonseca

**Prompt 2 Answer**:

Ana Trujillo

### Prompt 3

Among those customers who live in Germany, if you sort the customers in alphabetical order by name, which customer is fourth?

In [17]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT c.contact_name, c.country
FROM customers AS c
WHERE c.country = 'Germany'
ORDER BY c.contact_name

"""

df = pd.read_sql_query(sql, engine)
df.head()

contact_name  country
0      Alexander Feuer  Germany
1           Hanna Moos  Germany
2  Henriette Pfalzheim  Germany
3          Horst Kloss  Germany
4        Karin Josephs  Germany

**Prompt 3 Answer**:

Horst Kloss

### Prompt 4

From the `employees` table, what are the first names of the individuals who have a birthdate in the 1960s?

In [18]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT e.first_name, e.birth_date
FROM employees AS e
WHERE e.birth_date BETWEEN '1960-01-01' and '1969-12-31'


"""

df = pd.read_sql_query(sql, engine)
df.head()

first_name  birth_date
0      Janet  1963-08-30
1    Michael  1963-07-02
2     Robert  1960-05-29
3       Anne  1966-01-27

**Prompt 4 Answer**:

Janet, Michael, Robert, Anne

### Prompt 5

In `order_details` table of the Northwind database, what is the total number of units sold for products 2, 3, and 4? 
> You should give us three numbers - one for each `ProductID`.

In [19]:
sql = """
Select od.product_id, SUM(od.quantity)
FROM order_details AS od
WHERE od.product_id BETWEEN '2' AND '4' 
GROUP BY od.product_id
ORDER BY od.product_id

"""

df = pd.read_sql_query(sql, engine)
df.head()

product_id   sum
0           2  1057
1           3   328
2           4   453

**Prompt 5 Answer**:

>Product 2 units sold = 1057
>
>Product 3 units sold = 328
>
>Product 4 units sold = 453

### Prompt 6

From the `order_details` table of the Northwind database, calculate the revenue per product. In your SQL output, rename the revenue column `Revenue` and sort your results from largest revenue per product to smallest revenue per product. Which `ProductID` has the fifth-highest revenue, and what is that revenue rounded to the nearest dollar? 
> Note that revenue for a given product should be equal to the total unit price of that product times the quantity. You do not need to worry about the discount here. If you haven't multiplied columns together before, [this StackOverflow question](https://stackoverflow.com/questions/7536996/multiplying-two-columns-in-sql-server/7537059) may be helpful.

In [20]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT od.product_id, od.quantity * od.unit_price AS Revenue
FROM order_details AS od
ORDER BY Revenue DESC

"""

df = pd.read_sql_query(sql, engine)
df.head()

product_id       revenue
0          38  15810.000000
1          38  15810.000000
2          38  10540.000153
3          38  10540.000153
4          38  10540.000000

**Prompt 6 Answer**:

> Product ID 38 has the 5th largest revenue at 10,540

### Prompt 7

From the `products` table of the Northwind database, show the `ProductName`, `UnitsInStock`, and `UnitsOnOrder` values for those products who have more units on order than units in stock. Among these results, which product has the most units on order?

In [25]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT p.product_name, p.units_in_stock, p.units_on_order
FROM products AS p
WHERE p.units_on_order > p.units_in_stock
ORDER BY p.units_on_order DESC


"""

df = pd.read_sql_query(sql, engine)
df.head()

product_name  units_in_stock  units_on_order
0  Louisiana Hot Spiced Okra               4             100
1  Wimmers gute Semmelknödel              22              80
2          Gorgonzola Telino               0              70
3                Rogede sild               5              70
4              Aniseed Syrup              13              70

**Prompt 7 Answer**:

> Among these results, Louisiana Hot Spiced Okra has the most units on order

### Prompt 8

From the `order_details` table, show the products that have an average discount of at least 8%. Among these results, which average discount is closest to 8%?

In [13]:
sql = """
SELECT od.product_id, AVG(od.discount)
FROM order_details AS od
GROUP BY od.product_id
HAVING AVG(od.discount) >= 0.08
ORDER BY AVG(od.discount)DESC

"""

pd.read_sql_query(sql, engine)


product_id       avg
0          48  0.108333
1           2  0.102273
2           9  0.100000
3          58  0.100000
4          25  0.086111
5          30  0.081250

**Prompt 8 Answer**:

> Among these results, product ID # 30 has an average discount closest to 8%

### Prompt 9

Return the `orders` table, but also include the first name, last name, and country of the employee assigned to each order. What is the first and last name of the employee assigned to `OrderID` 11077? (This is the last row in the data, provided that you do not sort.)

In [27]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT o.*, e.last_name, e.first_name, e.country 
FROM orders AS o
INNER JOIN employees AS e on e.employee_id = o.employee_id
"""

df = pd.read_sql_query(sql, engine)
df.tail()

order_id customer_id  employee_id  order_date required_date shipped_date  \
825     11073       PERIC            2  1998-05-05    1998-06-02         None   
826     11074       SIMOB            7  1998-05-06    1998-06-03         None   
827     11075       RICSU            8  1998-05-06    1998-06-03         None   
828     11076       BONAP            4  1998-05-06    1998-06-03         None   
829     11077       RATTC            1  1998-05-06    1998-06-03         None   

     ship_via  freight                   ship_name              ship_address  \
825         2    24.95   Pericles Comidas clásicas  Calle Dr. Jorge Cash 321   
826         2    18.44               Simons bistro              Vinbæltet 34   
827         2     6.19          Richter Supermarkt               Starenweg 5   
828         2    38.28                    Bon app'      12, rue des Bouchers   
829         2     8.53  Rattlesnake Canyon Grocery           2817 Milton Dr.   

       ship_city ship_region ship_postal_code ship_country last_name  \
825  México D.F.        None            05033       Mexico    Fuller   
826    Kobenhavn        None             1734      Denmark      King   
827       Genève        None             1204  Switzerland  Callahan   
828    Marseille        None            13008       France   Peacock   
829  Albuquerque          NM            87110          USA   Davolio   

    first_name country  
825     Andrew     USA  
826     Robert      UK  
827      Laura     USA  
828   Margaret     USA  
829      Nancy     USA

**Prompt 9 Answer**:

> First Name: Nancy
>
> Last Name: Davolio

### Prompt 10

Find the average amount of freight across all orders, broken out by country of *employee*. (**NOT `ShipCountry`**.) How much higher is the average freight from UK employees than the average freight from US employees?

> For a bonus, round the freight to the nearest whole number and rename the column `Average Freight`.

In [19]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT AVG(o.freight) , e.country
FROM orders AS o
INNER JOIN employees AS e on e.employee_id = o.employee_id
GROUP BY e.country
"""

pd.read_sql_query(sql, engine)

avg country
0  78.977143      UK
1  77.973284     USA

**Prompt 10 Answer**:

> USA = 77.973284
>
> UK = 78.977143
>
> Difference = 1.003859